qiskit_ibm_runtime/execution_span.py

```diff
@@ -0,0 +1,141 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Execution span classes."""
+
+from typing import Iterable, Union, overload, Tuple, Dict, List, Iterator, Sequence
+from datetime import datetime
+from dataclasses import dataclass
+
+
+SliceType = Tuple[int, int]
+
+
+@dataclass(frozen=True)
+class ExecutionSpan:
+    """Stores an execution time span for a subset of job data."""
+
+    start: datetime
+    """The start time of the span, in UTC."""
+
+    stop: datetime
+    """The stop time of the span, in UTC."""
+
+    data_slices: Dict[int, SliceType]
+    r"""Which data have dependence on this execution span.
+
+    Data from the primitives are array-based, with every field in a 
+    :class:`~PubResult`\s :class:`~.DataBin` sharing the same base shape.
+    Therefore, the format of this field is a mapping from pub indexes to 
+    the same format accepted by 
+    NumPy slicing, where each value indicates which slice of each field in the
+    data bin depend on raw data collected during this execution span.
+    """
+
+    @property
+    def duration(self) -> float:
+        """Return the duration"""
+        return (self.stop - self.start).total_seconds()
+
+    def contains_pub(self, pub_idx: Union[int, Iterable[int]]) -> bool:
+        """Returns if a pub is contained"""
+        pub_idx = {pub_idx} if isinstance(pub_idx, int) else set(pub_idx)
+        return not pub_idx.isdisjoint(self.data_slices)
+
+    def filter_by_pub(self, pub_idx: Union[int, Iterable[int]]) -> "ExecutionSpan":
+        """Returns an ExecutionSpan filtered by pub"""
```

Comment:
```suggestion
        """Return a new span whose slices are filtered to the provided indices.
        
        For example, if this span contains slice information for pubs with indices 1,3,4 and `[1,4]` is provided,
        then the span returned by this method will contain slice information for only those two indices, but
        be identical otherwise.
        
        Args:
            pub_idx: One or more pub indices from the original primitive call.
            
        Returns:
            A new filtered span.
        """
```